File: backend/app/api/routes/listeners.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import Dict, Any, Optional

from app.core.msf_client import msf_client
from app.api.routes.auth import get_current_active_user

router = APIRouter()


class HandlerCreateRequest(BaseModel):
    payload: str
    lhost: str
    lport: int
    options: Optional[Dict[str, Any]] = {}


class JobInfo(BaseModel):
    job_id: str
    name: str
    start_time: Optional[int] = None


@router.get("/jobs")
async def list_jobs(user=Depends(get_current_active_user)):
    """List all running jobs (handlers, exploits, etc.)."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        jobs = await msf_client.list_jobs()
        result = []
        for job_id, name in jobs.items():
            result.append({
                "id": job_id,
                "name": name
            })
        return {"jobs": result, "count": len(result)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/jobs/{job_id}")
async def get_job_info(job_id: str, user=Depends(get_current_active_user)):
    """Get detailed information about a job."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        info = await msf_client.get_job_info(job_id)
        return info
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/jobs/{job_id}")
async def kill_job(job_id: str, user=Depends(get_current_active_user)):
    """Kill a running job."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        result = await msf_client.kill_job(job_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/handler")
async def create_handler(request: HandlerCreateRequest, user=Depends(get_current_active_user)):
    """Create a new multi/handler listener."""
    if not msf_client.connected:
        raise HTTPException(status_code=503, detail="Metasploit RPC not connected")

    try:
        options = {
            "LHOST": request.lhost,
            "LPORT": request.lport,
            "ExitOnSession": False,
            **request.options
        }

        result = await msf_client.create_handler(request.payload, options)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/payloads")
async def list_common_payloads(user=Depends(get_current_active_user)):
    """List commonly used payloads for handlers."""
    common_payloads = [
        {
            "name": "windows/meterpreter/reverse_tcp",
            "platform": "Windows",
            "arch": "x86",
            "type": "Meterpreter",
            "staged": True
        },
        {
            "name": "windows/x64/meterpreter/reverse_tcp",
            "platform": "Windows",
            "arch": "x64",
            "type": "Meterpreter",
            "staged": True
        },
        {
            "name": "windows/meterpreter_reverse_tcp",
            "platform": "Windows",
            "arch": "x86",
            "type": "Meterpreter",
            "staged": False
        },
        {
            "name": "windows/x64/meterpreter_reverse_tcp",
            "platform": "Windows",
            "arch": "x64",
            "type": "Meterpreter",
            "staged": False
        },
        {
            "name": "linux/x86/meterpreter/reverse_tcp",
            "platform": "Linux",
            "arch": "x86",
            "type": "Meterpreter",
            "staged": True
        },
        {
            "name": "linux/x64/meterpreter/reverse_tcp",
            "platform": "Linux",
            "arch": "x64",
            "type": "Meterpreter",
            "staged": True
        },
        {
            "name": "python/meterpreter/reverse_tcp",
            "platform": "Multi",
            "arch": "python",
            "type": "Meterpreter",
            "staged": True
        },
        {
            "name": "java/meterpreter/reverse_tcp",
            "platform": "Multi",
            "arch": "java",
            "type": "Meterpreter",
            "staged": True
        },
        {
            "name": "php/meterpreter/reverse_tcp",
            "platform": "Multi",
            "arch": "php",
            "type": "Meterpreter",
            "staged": True
        },
        {
            "name": "cmd/unix/reverse_bash",
            "platform": "Unix",
            "arch": "cmd",
            "type": "Shell",
            "staged": False
        },
        {
            "name": "windows/shell/reverse_tcp",
            "platform": "Windows",
            "arch": "x86",
            "type": "Shell",
            "staged": True
        },
        {
            "name": "generic/shell_reverse_tcp",
            "platform": "Multi",
            "arch": "multi",
            "type": "Shell",
            "staged": False
        },
    ]
    return {"payloads": common_payloads}

File: backend_ruby/standalone/routes/listeners.rb
# frozen_string_literal: true

require 'sinatra/base'

module MsfGui
  module Routes
    class Listeners < Sinatra::Base
      helpers Helpers::Response
      helpers Helpers::Validators

      configure do
        set :show_exceptions, false
      end

      COMMON_PAYLOADS = [
        {
          name: 'Windows Meterpreter Reverse TCP',
          payload: 'windows/meterpreter/reverse_tcp',
          platform: 'windows'
        },
        {
          name: 'Windows Meterpreter Reverse HTTPS',
          payload: 'windows/meterpreter/reverse_https',
          platform: 'windows'
        },
        {
          name: 'Windows x64 Meterpreter Reverse TCP',
          payload: 'windows/x64/meterpreter/reverse_tcp',
          platform: 'windows'
        },
        {
          name: 'Windows x64 Meterpreter Reverse HTTPS',
          payload: 'windows/x64/meterpreter/reverse_https',
          platform: 'windows'
        },
        {
          name: 'Linux Meterpreter Reverse TCP',
          payload: 'linux/x64/meterpreter/reverse_tcp',
          platform: 'linux'
        },
        {
          name: 'Linux Meterpreter Reverse HTTPS',
          payload: 'linux/x64/meterpreter/reverse_https',
          platform: 'linux'
        },
        {
          name: 'Android Meterpreter Reverse TCP',
          payload: 'android/meterpreter/reverse_tcp',
          platform: 'android'
        },
        {
          name: 'Android Meterpreter Reverse HTTPS',
          payload: 'android/meterpreter/reverse_https',
          platform: 'android'
        },
        {
          name: 'macOS Meterpreter Reverse TCP',
          payload: 'osx/x64/meterpreter/reverse_tcp',
          platform: 'macos'
        },
        {
          name: 'PHP Meterpreter Reverse TCP',
          payload: 'php/meterpreter/reverse_tcp',
          platform: 'php'
        },
        {
          name: 'Python Meterpreter Reverse TCP',
          payload: 'python/meterpreter/reverse_tcp',
          platform: 'python'
        },
        {
          name: 'Java Meterpreter Reverse TCP',
          payload: 'java/meterpreter/reverse_tcp',
          platform: 'java'
        },
        {
          name: 'Generic Shell Reverse TCP',
          payload: 'generic/shell_reverse_tcp',
          platform: 'multi'
        }
      ].freeze

      # List running handlers/jobs
      get '/jobs' do
        jobs = MsfGui.msf_client.list_jobs

        result = jobs.map do |id, name|
          info = MsfGui.msf_client.get_job_info(id)
          {
            id: id,
            name: name,
            start_time: info['start_time'],
            datastore: info['datastore']
          }
        end

        json_response(result)
      end

      # Get job details
      get '/jobs/:job_id' do
        info = MsfGui.msf_client.get_job_info(params[:job_id])

        if info.nil? || info.empty?
          json_error('Job not found', status: 404)
        end

        json_response({
          id: params[:job_id],
          name: info['name'],
          start_time: info['start_time'],
          datastore: info['datastore'],
          uripath: info['uripath']
        })
      end

      # Kill job
      delete '/jobs/:job_id' do
        result = MsfGui.msf_client.kill_job(params[:job_id])

        json_response({
          success: true,
          message: 'Job terminated',
          result: result
        })
      end

      # Create handler
      post '/handler' do
        body = parse_json_body

        payload = body[:payload] || body['payload']
        lhost = body[:lhost] || body['lhost'] || body['LHOST']
        lport = body[:lport] || body['lport'] || body['LPORT'] || 4444

        unless payload
          json_error('Payload is required', status: 422)
        end

        unless lhost
          json_error('LHOST is required', status: 422)
        end

        options = {
          'LHOST' => lhost,
          'LPORT' => lport.to_s
        }

        # Add any extra options
        extra_options = body[:options] || body['options'] || {}
        extra_options.each { |k, v| options[k.to_s.upcase] = v.to_s }

        result = MsfGui.msf_client.create_handler(payload, options)

        if result['error']
          json_error(result['error_message'] || 'Failed to create handler', status: 500)
        end

        json_response({
          success: true,
          job_id: result['job_id'],
          uuid: result['uuid'],
          message: "Handler started for #{payload}",
          payload: payload,
          lhost: lhost,
          lport: lport
        }, status: 201)
      end

      # Get common payloads for handlers
      get '/payloads' do
        json_response(COMMON_PAYLOADS)
      end

      # Quick start handler from template
      post '/quick' do
        body = parse_json_body

        template_index = (body[:template] || body['template'] || 0).to_i
        lhost = body[:lhost] || body['lhost']
        lport = body[:lport] || body['lport'] || 4444

        unless lhost
          json_error('LHOST is required', status: 422)
        end

        if template_index < 0 || template_index >= COMMON_PAYLOADS.size
          json_error("Invalid template index. Valid range: 0-#{COMMON_PAYLOADS.size - 1}", status: 400)
        end

        template = COMMON_PAYLOADS[template_index]
        payload = template[:payload]

        options = {
          'LHOST' => lhost,
          'LPORT' => lport.to_s
        }

        result = MsfGui.msf_client.create_handler(payload, options)

        if result['error']
          json_error(result['error_message'] || 'Failed to create handler', status: 500)
        end

        json_response({
          success: true,
          job_id: result['job_id'],
          template: template[:name],
          payload: payload,
          lhost: lhost,
          lport: lport
        }, status: 201)
      end
    end
  end
end
